### PR TITLE
HZ-397 - TO_TIMESTAMP_TZ SQL function

### PR DESCRIPTION
**Pull request by ivanthescientist**
Adds support for `TO_TIMESTAMP_TZ`  SQL function which converts arbitrary `BIGINT` values to `TIMESTAMP_WITH_TIMEZONE`. 

Example query: 
```
SELECT TO_TIMESTAMP_TZ(v) FROM TABLE (generate_series(1,3))
```

Example result: 
```
1970-01-01T03:00:00.001+03:00
1970-01-01T03:00:00.002+03:00
1970-01-01T03:00:00.003+03:00
```

### CODE REVIEW

**Comment by ivanthescientist** [comment]
run-lab-run

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Test
+    public void testColumn() {
+        checkColumn(1L, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(1L, ChronoUnit.SECONDS));
+        checkColumn(SECONDS, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(SECONDS, ChronoUnit.SECONDS));
+        checkColumn(MILLISECONDS, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(MILLISECONDS, ChronoUnit.MILLIS));
+        checkColumn(MICROSECONDS, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(MICROSECONDS, ChronoUnit.MICROS));
+        checkColumn(NANOSECONDS, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(NANOSECONDS, ChronoUnit.NANOS));
+        checkFailure("'1'", SqlErrorCode.PARSING, signatureError(SqlColumnType.VARCHAR));
+        checkFailure("1.0", SqlErrorCode.PARSING, signatureError(SqlColumnType.DECIMAL));
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzIntegrationTest.java:53]: These two should be in `testLiteral()` :). We should actually test with columns of unsupported types.

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Test
+    public void testLiteral() {
+        put(0);
+        checkLiteral(1L, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(1L, ChronoUnit.SECONDS));
+        checkLiteral(SECONDS, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(SECONDS, ChronoUnit.SECONDS));
+        checkLiteral(MILLISECONDS, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(MILLISECONDS, ChronoUnit.MILLIS));
+        checkLiteral(MICROSECONDS, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(MICROSECONDS, ChronoUnit.MICROS));
+        checkLiteral(NANOSECONDS, TIMESTAMP_WITH_TIME_ZONE, fromEpoch(NANOSECONDS, ChronoUnit.NANOS));
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzIntegrationTest.java:69]: We should also test 0 and negative values. And Long.MAX_VALUE and MIN_VALUE. And MAX_VALUE+1 (that should behave as DECIMAL)

**Comment by ivanthescientist** [comment]
run-lab-run

**Comment by ivanthescientist** [comment]
run-lab-run

**Comment by ivanthescientist** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Override
+    public QueryDataType getType() {
+        return QueryDataType.TIMESTAMP_WITH_TZ_OFFSET_DATE_TIME;
+    }
+
+    private TemporalUnit getChronoUnit(final long unixTimestamp) {
+        // -Long.MIN_VALUE returns Long.MIN_VALUE, probably because an absolute value of that is Long.MAX_VALUE + 1.
+        final long absTimestamp = unixTimestamp == Long.MIN_VALUE
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java:79]: Actually this is no longer the Snowflake semantics. And I think their semantics makes more sense. If you work with epoch-seconds, every value from -999999 to 9999 will work for you. If you work with epoch-millis or a finer unit, values >=1971-01-01T00:00 UTC will work. With ABS, you can't use values before 1970 at all.

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java (hunk 3 above)
Comment (ivanthescientist) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java:79]: So this `ABS` is here only to judge which unit is to be used with the actual computation (L67), so this should still allow to work with negative timestamps.  

I think their semantics to treat any negative timestamp as in seconds is more of a placeholder/temporary solution (judging by that caution section in docs - https://docs.snowflake.com/en/sql-reference/functions/to_timestamp.html#usage-notes). I think it would be worth keeping this logic here, wdyt? 

Maybe a better solution is to provide a second optional argument for this function which would allow overriding the timeunit?

Comment (viliam-durina) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java:79]: As I explained, I think it's actually better the way they have it now, not using ABS. For epoch-millis, the usable range is after 1971. It seems weird to have some millisecond support before 1969. Same for micros and nanos. 
When you're working with epoch-seconds, with ABS your year range is actually ~970..2970. Without ABS, the range is "unlimited" from below...

Practically it's the same, but from theoretical point of view, not using ABS makes more sense. And we'll at least be compatible with someone :)

**Comment by ivanthescientist** [comment]
run-lab-run

**Review by viliam-durina — APPROVED** [review]

**Review by hakanmemisoglu — APPROVED** [review]
Thanks @ivanthescientist. 
Can you just put a small comment on how this function interprets its argument based on different range?

**Comment by ivanthescientist** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java [hunk 4]
```diff
@@ -74,30 +74,47 @@ public QueryDataType getType() {
...
+     * </ol>
      *
-     * Note that this also imposes limits on the possible max correctly interpreted timestamp value,
-     * that would be otherwise correctly interpreted as in Seconds/Milliseconds/Microseconds.
-     * For example if this function is called on a unixTimestamp of 31_536_000_001 which in seconds
-     * corresponds to datetime of 2969-05-03 00:00:01, this function would instead interpret it
-     * as 1971-01-01 00:00:00.001. This is also true for maximum date expressed in milliseconds and microseconds.
-     * Additionally this of course also imposes the min date possible to be interpreted in Milliseconds,
-     * Microseconds and Nanoseconds.
+     * <p>Note that this also imposes limits on the possible max
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java:91]: The starting `<p>` should be on a separate line:
```suggestion
     * <p>
     * Note that this also imposes limits on the possible max
```

File: hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java [hunk 5]
```diff
@@ -74,30 +74,47 @@ public QueryDataType getType() {
...
+     * correctly interpreted timestamp value, that would be otherwise
+     * correctly interpreted as in seconds/milliseconds/microseconds.
+     * For example if this function is called on a unixTimestamp of
+     * 31_536_000_001 which in seconds corresponds to datetime of
+     * 2969-05-03 00:00:01, this function would instead interpret it
+     * as 1971-01-01 00:00:00.001. This is also true for maximum date
+     * expressed in milliseconds and microseconds. This of course also
+     * imposes the min date possible to be interpreted in milliseconds,
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java:99]: This is the longest line and has 64 characters, the max should be 72. It can be wrapped little longer.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java [hunk 6]
```diff
@@ -74,30 +74,47 @@ public QueryDataType getType() {
...
+     * microseconds and nanoseconds. Additionally any value less
+     * than 0 is treated as seconds regardless of the magnitude.</p>
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java:101]: We don't use the ending `</p>` or `</li>` - it's optional in html and looks better when read in plain text:
```suggestion
     * than 0 is treated as seconds regardless of the magnitude.
```

**Review by viliam-durina — CHANGES_REQUESTED** [review]

**Comment by ivanthescientist** [comment]
run-lab-run

**Comment by ivanthescientist** [comment]
run-lab-run

**Comment by ivanthescientist** [comment]
run-lab-run

**Comment by ivanthescientist** [comment]
run-lab-run

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        checkLiteral("-10000", TIMESTAMP_WITH_TIME_ZONE, fromEpoch(-10000L, ChronoUnit.SECONDS));
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzIntegrationTest.java:97]: We should have a test with a large enough negative value that would fall into the milliseconds range, if we did ABS.

**Review by viliam-durina — APPROVED** [review]
